### PR TITLE
Fix/5685 missing set username popup

### PR DESCRIPTION
**Pull request by wwalden**
#5685

## Checkpoints

**For developers**

- [x] If relevant, manual tests have been done to ensure the stability of the whole application and that the involved feature works
- [x] The original issue is up to date w.r.t the latest discussions and contains a short summary of the implemented solution
- [x] Automated tests covering the feature or the fix, have been written, deemed irrelevant (give the reason), or an issue has been created to implement the test (give the link)
- [x] Make sure that:
  - if there are modifications in the Tracim configuration files (eg. `development.ini`), they are documented in `backend/doc/setting.md`
  - any migration process required for existing instances is documented
  - relevant people for these changes are notified
- [x] Original authors of the features included in a multi-feature branch (maintenance fixes -> develop, security fixes -> develop, …) should be part of the reviewers, especially if you encountered merge conflicts.

**For code reviewers**

- [x] The code is clear enough
- [x] If there are FIXMEs in the code, related issues are mentioned in the FIXME
- [x] If there are TODOs, NOTEs or HACKs in code, the date and the developer initials are present

**For testers**

- [x] Manual, quality tests have been done

### CODE REVIEW

File: frontend/src/container/Tracim.jsx [hunk 1]
```diff
@@ -752,6 +753,10 @@ export class Tracim extends React.Component {
           />
         )}
 
+        {((!this.props.user.username && Cookies.get(COOKIE_FRONTEND.SHOW_USERNAME_POPUP) === 'true' && this.props.user.config.display_username_popup !== 'false') &&
+          <CardPopupUsername />
+        )}
+
```
Comment (gbonaspetti) [inline frontend/src/container/Tracim.jsx:759]:
```suggestion
        {((!props.user.username && Cookies.get(COOKIE_FRONTEND.SHOW_USERNAME_POPUP) === 'true' && props.user.config.display_username_popup !== 'false') &&
          <CardPopupUsername />
        )}

```



The hard coded strings 'true' and 'false' can't be consts or real booleans ?

File: frontend/src/container/CardPopupUsername.jsx [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+          <input
+            className='homepage__usernamePopup__body__input form-control'
+            type='text'
+            placeholder={props.t('Your username')}
+            value={this.state.newUsername}
+            onChange={this.handleChangeNewUsername}
+            data-cy='usernamePopup_username'
+          />
+
+          {!this.state.isUsernameValid && (
```
Comment (gbonaspetti) [inline frontend/src/container/CardPopupUsername.jsx:146]:
```suggestion
          {!state.isUsernameValid && (
```

File: frontend/src/container/CardPopupUsername.jsx [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+            <div className='homepage__usernamePopup__errorMsg'>
+              <i className='homepage__usernamePopup__errorIcon fas fa-times' />
+              {this.state.usernameInvalidMsg}
```
Comment (gbonaspetti) [inline frontend/src/container/CardPopupUsername.jsx:149]:
```suggestion
              {state.usernameInvalidMsg}
```

File: frontend/src/container/CardPopupUsername.jsx [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+            </div>
+          )}
+
+          {this.state.isUsernameValid && (
```
Comment (gbonaspetti) [inline frontend/src/container/CardPopupUsername.jsx:153]:
```suggestion
          {state.isUsernameValid && (
```

File: frontend/src/container/CardPopupUsername.jsx [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+            <div className='homepage__usernamePopup__infoMsg'>
+              {props.t('Allowed characters: {{allowedCharactersUsername}}', { allowedCharactersUsername: ALLOWED_CHARACTERS_USERNAME })}
+            </div>
+          )}
+
+          {this.state.newUsername !== '' && (
```
Comment (gbonaspetti) [inline frontend/src/container/CardPopupUsername.jsx:159]:
```suggestion
          {state.newUsername !== '' && (
```

File: frontend/src/container/CardPopupUsername.jsx [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+            <>
+              <div className='homepage__usernamePopup__body__msg'>
+                {props.t('Please confirm your password:')}
+              </div>
+
+              <input
+                className='homepage__usernamePopup__body__input form-control'
+                type='password'
+                placeholder={props.t('Password')}
+                value={this.state.password}
```
Comment (gbonaspetti) [inline frontend/src/container/CardPopupUsername.jsx:169]:
```suggestion
                value={state.password}
```

File: frontend/src/container/CardPopupUsername.jsx [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+          return false
+        default:
+          props.dispatch(newFlashMessage(props.t('Error while changing username'), 'warning'))
+          return false
+      }
+    }
+    this.handleClickCloseUsernamePopup()
+  }
+
+  handleChangeNewUsername = async e => {
```
Comment (gbonaspetti) [inline frontend/src/container/CardPopupUsername.jsx:93]: Why async ?

**Review by gbonaspetti — CHANGES_REQUESTED** [review]

File: functionnal_tests/cypress/integration/login/username_popup_spec.js [hunk 8]
```diff
@@ -98,13 +95,15 @@ describe('Login', function () {
...
       cy.get(confirmButton).click()
+      cy.get('.flashmessage__container__content__text__title').contains('Error').should('not.exist')
       cy.get('.flashmessage').contains('Invalid password')
     })
 
     it('should be able to set username', function () {
       cy.get(usernameInput).type(newUsername)
       cy.get(passwordInput).type(userWithoutUsername.password)
       cy.get(confirmButton).click()
+      cy.get('.flashmessage__container__content__text__title').contains('Error').should('not.exist')
```
Comment (gbonaspetti) [inline functionnal_tests/cypress/integration/login/username_popup_spec.js:106]:
```suggestion
      cy.contains('.flashmessage__container__content__text__title', 'Error').should('not.exist')
      cy.get('.flashmessage').contains('Invalid password')
    })

    it('should be able to set username', function () {
      cy.get(usernameInput).type(newUsername)
      cy.get(passwordInput).type(userWithoutUsername.password)
      cy.get(confirmButton).click()
      cy.contains('.flashmessage__container__content__text__title', 'Error').should('not.exist')
```

**Review by wwalden — COMMENTED** [review]
ok

**Review by wwalden — COMMENTED** [review]
ok

**Review by wwalden — COMMENTED** [review]
ok

**Review by wwalden — COMMENTED** [review]
ok

**Review by wwalden — COMMENTED** [review]
ok

**Review by gbonaspetti — APPROVED** [review]

**Comment by gbonaspetti** [comment]
When the user chose to not show the popup again, in the next time he log in, we have the popup that appears for some milliseconds and disappears after (see gif bellow). I think it is a small bug and we can merge with it. 
It would be nice to make an issue about it.
username



Tested on branch

**Pull request merged**